File: set_finder/poc/set_finder/shape_detector.py
import cv2

from cards import Shape


class ShapeDetector:
    def __init__(self, cardPreprocessor):
        self._shape = Shape.UNKNOWN
        contours = cardPreprocessor.getContours()

        biggestContour = max(contours, key=abs and cv2.contourArea)
        contourArea = abs(cv2.contourArea(biggestContour, True))
        _, _, w, h = cv2.boundingRect(biggestContour)

        diffArea = w * h - contourArea
        if diffArea >= 1800:
            self._shape = Shape.DIAMOND
        elif contourArea >= 3600 and diffArea <= 1000:
            self._shape = Shape.OVAL
        else:
            self._shape = Shape.SQUIGGLE

    def getShape(self):
        return self._shape

File: set_finder/core/src/detector/shape_detector.cpp

#include "detector/shape_detector.h"
#include "utils/logger.h"

#include <algorithm>
#include <opencv2/imgproc.hpp>

namespace
{
constexpr uint32_t DIAMOND_DIFF_AREA_TH = 1800;
constexpr uint32_t OVAL_DIFF_AREA_TH = 1000;
constexpr uint32_t OVAL_AREA_TH = 3600;
} // namespace

Feature ShapeDetector::detect_feature(const CardPreprocessor& preprocessor) const
{
    const auto& contours = preprocessor.get_contours();
    if(contours.empty())
    {
        return Shape::UNKNOWN;
    }

    const auto biggest_contour = std::max_element(
        std::begin(contours), std::end(contours), [](const auto& contour1, const auto& contour2) {
            return std::abs(cv::contourArea(contour1)) > std::abs(cv::contourArea(contour2));
        });

    const auto biggest_area = abs(cv::contourArea(*biggest_contour, true));
    const auto bounding_rect = cv::boundingRect(*biggest_contour);
    const auto diff_area = bounding_rect.width * bounding_rect.height - biggest_area;

    if(diff_area >= DIAMOND_DIFF_AREA_TH)
    {
        return Shape::DIAMOND;
    }
    else if(biggest_area >= OVAL_AREA_TH and diff_area <= OVAL_DIFF_AREA_TH)
    {
        return Shape::OVAL;
    }

    return Shape::SQUIGGLE;
}
